framework/src/main/java/org/checkerframework/framework/util/ContractsUtils.java

```diff
@@ -417,19 +418,61 @@ private AnnotationMirror getAnnotationMirrorOfMetaAnnotation(
             }
         }
 
-        // Check type-system specific annotations.
-        Class<PostconditionAnnotation> metaAnnotation = PostconditionAnnotation.class;
+        // Check for type-system specific annotations.
         List<Pair<AnnotationMirror, AnnotationMirror>> declAnnotations =
-                factory.getDeclAnnotationWithMetaAnnotation(methodElement, metaAnnotation);
+                factory.getDeclAnnotationWithMetaAnnotation(
+                        methodElement, PostconditionAnnotation.class);
         for (Pair<AnnotationMirror, AnnotationMirror> r : declAnnotations) {
             AnnotationMirror anno = r.first;
             AnnotationMirror metaAnno = r.second;
-            AnnotationMirror postcondAnno = getAnnotationMirrorOfMetaAnnotation(metaAnno, anno);
+
+            result.addAll(getPostConditionAnnotations(anno, metaAnno));
+        }
+
+        return result;
+    }
+
+    /**
+     * Returns a collection of postcondition annotations that can consist of
+     * {@code @EnsuresNonNull},{@code @EnsuresKeyFor},{@code @EnsuresLTLengthOf},
+     * {@code @EnsuresLockHeld} or any other postcondition annotation developed in the future.
+     *
+     * @param anno wrapper annotation if multiple annotations are present, otherwise single
+     *     annotation
+     * @param metaAnno postcondition meta-annotation that indicates that an annotation is a
+     *     postcondition annotation
+     */
+    private Set<Postcondition> getPostConditionAnnotations(
+            AnnotationMirror anno, AnnotationMirror metaAnno) {
+
+        @SuppressWarnings("unchecked")
+        List<AnnotationValue> elementvalue =
+                AnnotationUtils.getElementValue(anno, "value", List.class, false);
+
+        Set<Postcondition> result = new LinkedHashSet<>();
```

Comment:
You have not answered my question about the relationship between these two variables.
Is it the case that one of them is always empty?  (I think so.)

More generally, the flow of control in this method is very confusing.  It would be better to do a single test, at the beginning of the method, and put all the code for handling a wrapper in one branch and all the code for handling a single annotation in the other branch.

Reply:
Yes, you are right. Divided it in two blocks.